pytorch_translate/char_rnn_encoder.py

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python3
+
+import logging
+import math
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.utils.rnn import pack_padded_sequence, PackedSequence, pad_packed_sequence
+
+from pytorch_translate import rnn  # noqa
+
+logger = logging.getLogger(__name__)
+
+
+def add_args(parser):
+    parser.add_argument(
+        "--char-rnn",
+        action="store_true",
+        default=False,
+        help="Assumes input is delimiter-separated character tokens. "
+        "Encodes word representations using bi-LSTM over char inputs.",
+    )
+    parser.add_argument(
+        "--char-rnn-units",
+        type=int,
+        default=256,
+        metavar="N",
+        help=("Number of units for Character LSTM."),
+    )
+
+    parser.add_argument(
+        "--char-rnn-layers",
+        type=int,
+        default=2,
+        metavar="N",
+        help=("Number of Character LSTM layers."),
+    )
+    parser.add_argument(
+        "--word-delimiter", type=str, default="@SPACE", help=("Token separating words.")
+    )
+
+
+def set_arg_defaults(args):
+    if hasattr(args, "char_rnn_params"):
+        return args.char_rnn_params
+
+    args.char_rnn_params = None
+    if args.char_rnn:
+        args.char_rnn_params = {
+            "char_rnn_units": args.char_rnn_units,
+            "char_rnn_layers": args.char_rnn_layers,
+            "word_delimiter": args.word_delimiter,
+        }
+
+
+class DelimiterSplit(nn.Module):
+    """
+    nn.Module which takes batched sequence input where the tokens are assumed
+    to represent characters with a specified delimiter separating words, and
+    returns the same indices split into words.
+
+    Inputs:
+    src_tokens (batch_size, max_length): character indices
+    src_lengths (batch_size): lengths in total characters including delimiters
+
+    Outputs:
+    padded_char_inds (max_word_length, total_words)
+    word_lenths (total_words,)
+    words_per_sent (batch_size,)
+    """
+
+    def __init__(self, dictionary, word_delimiter="@SPACE"):
+        super().__init__()
+        self.dictionary = dictionary
+        self.padding_idx = dictionary.pad()
+        self.word_delim_index = self.dictionary.index(word_delimiter)
+        if self.word_delim_index == self.dictionary.unk():
+            logger.warning(f"Word delimiter {word_delimiter} not in source dictionary!")
```

Comment:
Does it make sense to continue in this case? Would it make sense to throw an exception?